NonSucking.Framework.Serialization.Advanced/BinarySpanWriter.cs

```diff
@@ -0,0 +1,265 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+using NonSucking.Framework.Serialization;
+
+namespace NonSucking.Framework.Serialization
+{
+    class ThrowHelper
+    {
+#if NETSTANDARD2_1_OR_GREATER
+        [DoesNotReturn]
+#endif
+        public static void ThrowIndexOutOfRange() => throw new IndexOutOfRangeException();
+    }
+
+    public ref struct BinarySpanWriter
+    {
+        private readonly Span<byte> buffer;
+        private int offset;
+
+        private readonly Encoding encoding;
+
+        public BinarySpanWriter(Span<byte> buffer)
+            : this(buffer, Encoding.UTF8)
+        {
+        }
+
+        public BinarySpanWriter(Span<byte> buffer, Encoding encoding)
+        {
+            this.buffer = buffer;
+            offset = 0;
+            this.encoding = encoding;
+        }
+
+        public void Write(bool value)
+        {
+            Write((byte)(value ? 1 : 0));
+        }
+
+
+        public void Write(byte value)
+        {
+            buffer[offset++] = value;
+        }
+
+        public void Write(byte[] buffer)
+        {
+            Write(new ReadOnlySpan<byte>(buffer));
+        }
+
+        public void Write(byte[] buffer, int index, int count)
+        {
+            Write(new ReadOnlySpan<byte>(buffer, index, count));
+        }
+
+        public void Write(char ch)
+        {
+#if NETSTANDARD2_1_OR_GREATER
+            ReadOnlySpan<char> chars = stackalloc char[] { ch};
+            var b = encoding.GetByteCount(chars);
+            if (b != encoding.GetBytes(chars, buffer.Slice(offset)))
+                ThrowHelper.ThrowIndexOutOfRange();
+#else
+            var chars = new [] { ch };
+            var b = encoding.GetByteCount(chars);
+            var res = encoding.GetBytes(chars);
+            if (b != res.Length)
+                ThrowHelper.ThrowIndexOutOfRange();
+            res.CopyTo(buffer.Slice(offset));
+#endif
+        }
+
+        public void Write(char[] chars)
+        {
+            Write(new ReadOnlySpan<char>(chars));
+        }
+
+        public void Write(char[] chars, int index, int count)
+        {
+            Write(new ReadOnlySpan<char>(chars, index, count));
+        }
+
+        public void Write(decimal value)
+        {
+#if NET5_OR_GREATER
+            Span<int> buffer = stackalloc int[sizeof(decimal) / sizeof(int)];
+            _ = decimal.TryGetBits(value, buffer, out _);
```

Comment:
You don't care about the bool, so you can just use the GetBits function instead
```suggestion
            _ = decimal.GetBits(value, buffer);
```